src/compute-client/src/controller/replicated.rs

```diff
@@ -447,12 +469,15 @@ where
                 //    information as a `FrontierUppers` response.
 
                 let old_upper_bound = entry.bounds.upper.clone();
-                if !entry.update(replica_id, upper) {
+                if !entry.update(replica_id, upper.clone()) {
```

Comment:
Nit: It seems we could avoid the `clone` by remembering only if the frontier is empty.